### PR TITLE
Add DATERANGE support

### PR DESCRIPTION
**Pull request by harisha-swaminathan**
Adds support for #EXT-X-DATERANGE

https://datatracker.ietf.org/doc/html/draft-pantos-http-live-streaming-23#section-4.3.2.7

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -631,6 +631,54 @@ export default class Parser extends Stream {
...
+              }
+              if (daterange.duration && daterange.endDate && (new Date(daterange.endDate) !== new Date(daterange.startDate) + daterange.duration)) {
+                this.trigger('warn', {
+                  message: 'EXT-X-DATERANGE with DURATION and END-DATE attribute must have END-DATE equal to START-DATE +  DURATION'
+                });
+              }
+
+              if (daterange && !this.manifest.dateTimeString) {
+                this.trigger('warn', {
+                  message: 'A playlist with EXT-X-DATERANGE tag must contain atleast one EXT-X-PROGRAM-DATE-TIME tag'
```
Comment (philjhale) [inline src/parser.js:679]:
```suggestion
                  message: 'A playlist with EXT-X-DATERANGE tag must contain at least one EXT-X-PROGRAM-DATE-TIME tag'
```

File: src/parse-stream.js [hunk 2]
```diff
@@ -567,6 +567,51 @@ export default class ParseStream extends Stream {
...
+          }
+        });
+
+        ['SCTE35-CMD', ' SCTE35-OUT', 'SCTE35-IN'].forEach(function(key) {
+          if (event.attributes.hasOwnProperty(key)) {
+            event.attributes[key] = parseInt(event.attributes[key], 16);
+          }
+        });
+
+        const clientAttributePattern = /^X-([A-Z]+-)+[A-Z]+$/;
```
Comment (philjhale) [inline src/parse-stream.js:599]: Is it worth adding some tests for these client attributes?

File: src/parse-stream.js [hunk 3]
```diff
@@ -567,6 +567,51 @@ export default class ParseStream extends Stream {
...
+            event.attributes[key] = new Date(event.attributes[key]);
+          }
+        });
+        ['DURATION', 'PLANNED-DURATION'].forEach(function(key) {
+          if (event.attributes.hasOwnProperty(key)) {
+            event.attributes[key] = parseFloat(event.attributes[key]);
+          }
+        });
+
+        ['SCTE35-CMD', ' SCTE35-OUT', 'SCTE35-IN'].forEach(function(key) {
```
Comment (philjhale) [inline src/parse-stream.js:593]: Is it worth adding tests for this attributes?

**Comment by philjhale** [comment]
We can also update [the documentation](https://github.com/videojs/m3u8-parser/blob/main/README.md?plain=1#L239) to say `#EXT-X-DATERANGE` is supported.

File: src/parse-stream.js [hunk 4]
```diff
@@ -567,6 +567,51 @@ export default class ParseStream extends Stream {
         this.trigger('data', event);
         return;
       }
+      match = (/^#EXT-X-DATERANGE:(.*)$/).exec(newLine);
```
Comment (philjhale) [inline src/parse-stream.js:570]: My understanding from reading the spec is that there could be multiple `DATERANGE` tags. Will this code support multiple tags?

Comment (harisha-swaminathan) [reply src/parse-stream.js:570]: Just pushed a change to support multiple tags!

File: src/parse-stream.js (hunk 2 above)
Comment (harisha-swaminathan) [reply src/parse-stream.js:599]: Done!

**Review by adrums86 — APPROVED** [review]
Great work, Harisha! This seems to line up with the spec well, yet provide enough resiliency through warning rather than erroring out on non-compliant DATERANGE tags. :shipit:

**Review by dzianis-dashkevich — APPROVED** [review]

**Comment by dzianis-dashkevich** [comment]
LGTM, just one comment

**Pull request merged**